src/ServiceControl.Audit/Infrastructure/RavenDB/Expiration/KnownEndpointsCleaner.cs

```diff
@@ -0,0 +1,100 @@
+﻿namespace ServiceControl.Audit.Infrastructure.RavenDB.Expiration
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.Linq;
+    using System.Threading;
+    using NServiceBus.Logging;
+    using Raven.Abstractions;
+    using Raven.Abstractions.Commands;
+    using Raven.Abstractions.Data;
+    using Raven.Database;
+
+    static class KnownEndpointsCleaner
+    {
+        public static void Clean(int deletionBatchSize, DocumentDatabase database, DateTime expiryThreshold, CancellationToken token)
+        {
+            var stopwatch = Stopwatch.StartNew();
+            var items = new List<ICommandData>(deletionBatchSize);
+
+            try
+            {
+                var query = new IndexQuery
```

Comment:
I first wanted to use something like

```
Operation operation = store
	.DatabaseCommands
	.DeleteByIndex(
		"Raven/DocumentsByEntityName",
		new IndexQuery
		{
			Query = "Tag:KnownEndpoints AND LastSeen:[* TO {expiryThreshold.Ticks}]"
		},
		new BulkOperationOptions
		{
			AllowStale = false
		});
```

but then I wasn't able to find delete by index on the document database structure and I realized in order to filter by LastSeen we require an index anyway

Reply:
Elsewhere in the app, we're making the assumption that the total number of known endpoints will generally be contained to < 1024. So why are we doing the select + delete ids in batch dance when we could do a DeleteByIndex using ExpiryKnownEndpointsIndex?

Reply:
Cleaning runs in a bundle, there I have only access to the Document Database type. Like written before I couldn't find a delete by index so I used the same approach as in the audit cleaner to use the batch size settings to chunk for convenience. Regarding the settings yeah we kind of hardcoded page size to be 1024 for now and cutt off the query but we also introduced paging in the other PR so I'm on the fence.

Do you know of a way of doing delete by query on the Document Database object by any chance?

Reply:
Ah, I forgot this was in a bundle. 🤦

Reply:
Have we considered just setting the Raven Expiration Date, sliding it forward for each new entry, and relying on the built-in retention bundle to do the job?

https://ravendb.net/docs/article-page/3.5/Csharp/server/bundles/expiration

Reply:
I dismissed Expiration for now because as of today Expiration is a business feature. Meaning if someone changes the expiration time it will affect all running cycles after the date has been set. This is also doable with Expiration but we would have to apply a set based patch then on the endpoints. I think this is likely coming up again as a discussion point when upgrading Raven so I figured I just stick to the current pattern